nntrainer/opencl/opencl_buffer.cpp

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Debadri Samaddar <s.debadri@samsung.com>
+ *
+ * @file    opencl_buffer.cpp
+ * @date    06 Feb 2024
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Debadri Samaddar <s.debadri@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   OpenCL wrapper for buffer usage
+ *
+ */
+
+#include "opencl_buffer.h"
+
+#include "opencl_loader.h"
+
+#include <nntrainer_log.h>
+
+namespace nntrainer::opencl {
+
+/**
+ * @brief Construct a new Buffer object
+ *
+ * @param context_manager reference
+ * @param size_in_bytes size of buffer
+ * @param read_only flag
+ * @param data data for the buffer
+ */
+Buffer::Buffer(ContextManager &context_manager, int size_in_bytes,
+               bool read_only, void *data) {
+  cl_context context = context_manager.GetContext();
+  cl_mem_flags flags = read_only ? CL_MEM_READ_ONLY : CL_MEM_READ_WRITE;
+  if (data) {
+    flags |= CL_MEM_COPY_HOST_PTR;
+  }
+  cl_int error_code;
+
+  // clCreateBuffer returns NULL with error code if fails
+  mem_buf_ = clCreateBuffer(context, flags, size_in_bytes, data, &error_code);
+  size_ = size_in_bytes;
+  if (!mem_buf_) {
+    size_ = 0;
+    ml_loge("Failed to allocate device memory (clCreateBuffer). OpenCL error "
+            "code: %d",
+            error_code);
+  }
+}
+
+/**
+ * @brief Copy constructor for buffer
+ *
+ * @param buffer
+ */
+Buffer::Buffer(Buffer &&buffer) :
+  mem_buf_(buffer.mem_buf_),
+  size_(buffer.size_) {
+  buffer.mem_buf_ = nullptr;
+  buffer.size_ = 0;
```

Comment:
could you explain why resetting the input buffer?

Reply:
This is a move constructor. To copy over the data and delete from the previous. Updating the doc in next commit.